### PR TITLE
Add API versioning and deprecation policy

### PR DESCRIPTION
**Pull request by varbanv**
**Description**

Changes proposed in this pull request:

- Document API versioning and deprecation information

**Related issue(s)**
https://github.com/kyma-project/kyma/issues/14737

### CODE REVIEW

File: docs/05-technical-reference/00-api-overview/README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+**Rule #3: An API version in a given track may not be deprecated in favor of a less stable API version.**
+
+  * GA API versions can replace beta and alpha API versions.
+  * Beta API versions can replace earlier beta and alpha API versions, but *may not* replace GA API versions.
+  * Alpha API versions can replace earlier alpha API versions, but *may not* replace GA or beta API versions.
+
+**Rule #4a: minimum API lifetime is determined by the API stability level**
+
+   * **GA API versions may be marked as deprecated, but must not be removed within a major version of a Kyma module**
+   * **Beta API versions must be supported for 9 months or 3 releases (whichever is longer) after deprecation**
```
Comment (pbochynski) [inline docs/05-technical-reference/00-api-overview/README.md:69]:
```suggestion
   * **Beta API versions must be supported for 6 months or 3 releases (whichever is longer) after deprecation**
```

**Review by pbochynski — CHANGES_REQUESTED** [review]
Majority of the content is copied from https://kubernetes.io/docs/reference/using-api/deprecation-policy
Please make it a quote, provide the source (or maybe just link it instead).

**Review by pbochynski — DISMISSED** [review]

File: docs/05-technical-reference/00-api-overview/README.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+### REST resources (aka API objects)
+
+Consider a hypothetical REST resource named Widget, which was present in API v1 in the above timeline, and which needs to be deprecated. We document and announce the deprecation in sync with release X+1. The Widget resource still exists in API version v1 (deprecated) but not in v2alpha1. The Widget resource continues to exist and function in releases up to and including X+8. Only in release X+9, when API v1 has aged out, does the Widget resource cease to exist, and the behavior get removed.
+
+### Fields of REST resources
+
+As with whole REST resources, an individual field which was present in API v1 must exist and function until API v1 is removed.  Unlike whole resources, the v2 APIs may choose a different representation for the field, as long as it can be round-tripped. For example a v1 field named "magnitude" which was deprecated might be named "deprecatedMagnitude" in API v2. When v1 is eventually removed, the deprecated field can be removed from v2.
```
Comment (IwonaLanger) [inline docs/05-technical-reference/00-api-overview/README.md:84]: The phrase "whole REST resources" is used a few times and I'd like to clarify if the meaning is all of REST resources meaning each of them, or each of them as a whole - shall we say - unit.

Comment (varbanv) [reply docs/05-technical-reference/00-api-overview/README.md:84]: It should be the second - so you can deprecate the whole REST resource, or just one or more fields.

File: docs/05-technical-reference/00-api-overview/README.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+These rules are enforced between official releases, not between arbitrary commits to main or release branches.
+
+**Rule #1: API elements may only be removed by incrementing the version of the API group.**
+
+Once an API element has been added to an API group at a particular version, it cannot be removed from that version or have its behavior significantly changed, regardless of the track.
+
+**Rule #2: API objects must be able to round-trip between API versions in a given release without information loss, with the exception of whole REST resources that do not exist in some versions.**
+
+For example, an object can be written as v1 and then read back as v2 and converted to v1, and the resulting v1 resource will be identical to the original.  The representation in v2 might be different from v1, but the system knows how to convert between them in both directions.  Additionally, any new field added in v2 must be able to round-trip to v1 and back, which means v1 might have to add an equivalent field or represent it as an annotation.
```
Comment (IwonaLanger) [inline docs/05-technical-reference/00-api-overview/README.md:60]:
```suggestion
For example, an object can be written as v1 and then read back as v2 and converted to v1, and the resulting v1 resource is identical to the original.  The representation in v2 might be different from v1, but the system knows how to convert between them in both directions.  Additionally, any new field added in v2 must be able to round-trip to v1 and back, which means v1 might have to add an equivalent field or represent it as an annotation.
```

File: docs/05-technical-reference/00-api-overview/README.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  * Beta API versions can replace earlier beta and alpha API versions, but *may not* replace GA API versions.
+  * Alpha API versions can replace earlier alpha API versions, but *may not* replace GA or beta API versions.
+
+**Rule #4a: minimum API lifetime is determined by the API stability level.**
+
+   * **GA API versions may be marked as deprecated, but must not be removed within a major version of a Kyma module**
+   * **Beta API versions must be supported for 6 months or 3 releases (whichever is longer) after deprecation**
+   * **Alpha API versions may be removed in any release without prior deprecation notice**
+
+**Rule #4b: The `preferred` API version and the `storage version` for a given group may not advance until after a release has been made that supports both the new version and the previous version**
```
Comment (IwonaLanger) [inline docs/05-technical-reference/00-api-overview/README.md:74]:
```suggestion
**Rule #4b: The `preferred` API version and the `storage version` for a given group may not advance until after a release has been made that supports both the new version and the previous version.**
```

**Review by IwonaLanger — APPROVED** [review]

**Pull request merged**